### ISSUE TITLE
SpriteSheet load complete fires before the associated image is loaded

### ISSUE THREAD

**Issue by stevenalanstark** [pixijs/pixi.js#5916]
Hello,

I am finding myself continually fighting with SpriteSheets and knowing when their associated image(s) is finished loading.

The core of the issue is that the loader must load the `.json` spritesheet file, parse and read it, look for `meta.image` in the `.json`, and then add that resource into the loader.

Well, the loader will often fire it's complete signal as soon as the `.json` file is complete, without waiting for the associated images.

This is not the case all the time, it only shows up now and then. Reload the page and things work again. Nothing is being cached either, so it's not a simple caching issue.

I'm wanting to be able to load multiple sprite sheets, their associated images, and all other sounds, fonts, etc in one batch so that I can use the complete handler to continue with the initialization of my game.

I have tried both the Signal cb function on `LOADER.load((loader, resources) => {})` as well as adding a listener to `LOADER.onProgress.add` and checking that `prog.progress >= 99.9` ( and `100` ) as well as using the `LOADER.onComplete.add`. 

All three approaches have the same result.

I had a suspicion that this might be caused by exceeding the default value of 10 for `LOADER.concurrency`, however setting this to `100` before the load call didn't change anything.

I posted this issue here and not in the Loader repo because this seems to be an issue with the SpriteSheet class more than with the loader itself, but I may be mistaken in that assumption. LMK if you'd like me to move this issue.

### Expected Behavior

loading a spritesheet .json file will block the loader's complete signal until the sprite sheet data is parsed, and determined if more assets need to be loaded before completing the load sequence.

### Current Behavior

loading a spritesheet .json will automatically load additional assets, but these assets do not count as far as the onComplete signal is concerned.

- **`pixi.js` version**: 5.0.4
- **Browser & Version**: Chrome
- **OS & Version**: OSX ( assume the issue exists on all OS )

**Comment by YuriyEmpty** [comment]
Faced the same problem with a loading resources for "pixi-spine" plugin. The problem occurs after callback loaded resource. The object itself is formed and loaded, but "spineData" is undefined. This problem occurs once approximately every ten attempt.

**Comment by eXponenta** [comment]
> Hello,
> 
> I am finding myself continually fighting with SpriteSheets and knowing when their associated image(s) is finished loading.
> 
> The core of the issue is that the loader must load the `.json` spritesheet file, parse and read it, look for `meta.image` in the `.json`, and then add that resource into the loader.
> 
> Well, the loader will often fire it's complete signal as soon as the `.json` file is complete, without waiting for the associated images.
> 
> This is not the case all the time, it only shows up now and then. Reload the page and things work again. Nothing is being cached either, so it's not a simple caching issue.
> 
> I'm wanting to be able to load multiple sprite sheets, their associated images, and all other sounds, fonts, etc in one batch so that I can use the complete handler to continue with the initialization of my game.
> 
> I have tried both the Signal cb function on `LOADER.load((loader, resources) => {})` as well as adding a listener to `LOADER.onProgress.add` and checking that `prog.progress >= 99.9` as well as using the `LOADER.onComplete.add`.
> 
> All three approaches have the same result.
> 
> I had a suspicion that this might be caused by exceeding the default value of 10 for `LOADER.concurrency`, however setting this to `100` before the load call didn't change anything.
> 
> I posted this issue here and not in the Loader repo because this seems to be an issue with the SpriteSheet class more than with the loader itself, but I may be mistaken in that assumption. LMK if you'd like me to move this issue.
> 
> ### Expected Behavior
> loading a spritesheet .json file will block the loader's complete signal until the sprite sheet data is parsed, and determined if more assets need to be loaded before completing the load sequence.
> 
> ### Current Behavior
> loading a spritesheet .json will automatically load additional assets, but these assets do not count as far as the onComplete signal is concerned.
> 
> * **`pixi.js` version**: 5.0.4
> * **Browser & Version**: Chrome
> * **OS & Version**: OSX ( assume the issue exists on all OS )

Could you create fiddle for bug representation on https://pixiplayground.com .

**Comment by ivanpopelyshev** [comment]
I know we need a demo, but maybe you already suspect something, @englercj ?

**Comment by stevenalanstark** [comment]
I thought about the need for a demo as well, but given the inconsistent nature of this issue I'm not sure how feasible that is. Do you have any suggestions for that? I guess I could use Cypress.io and run it until it errors...

I can tell you that I have been seeing this for the over a year. I keep thinking I have it fixed, then it pops up again. I've gone as far as to place arbitrary delays before continuing in order to allow the images to load, but that kind of an approach is unreliable.

The last think I can think of is to not let the spritesheet control the image loading, and instead control that myself. But I rather like how it can just handle that from the `.json`, which is very convenient as far as the art pipeline is concerned. The less work we need to do to integrate new art assets, the better.

**Comment by stevenalanstark** [comment]
Update: I found the issue was actually in a completely unexpected place; `Promise.all` was the culprit in my project. 

I wrap the complete signal in a Promise, and reference it in a Promise.all() call, along with initializing my models. 

Well I confirmed today that it was the Promise.all call that was resolving before it was supposed to. Honestly, I don't know how this is even possible, but removing the Promise.all in favour of a promise chain resolved the issue for me.

Closing this issue.
